### PR TITLE
feat: genkit generate synthetic database

### PR DESCRIPTION
**Pull request by aikozhaoz**
# Description

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [ ] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [ ] You are listed as the author in your notebook or README file.
  - [ ] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [ ] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [ ] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

**Review by holtskinner — CHANGES_REQUESTED** [review]
Thanks for the contribution! Please address any linter/spelling errors detected and answer the comments.

For any false-positive spelling errors, add them to `.github/actions/spelling/allow.txt`

I think the only two that are actual errors:

- commited → committed
- dogfood → dog food

File: genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+import { gemini15ProPreview, vertexAI } from '@genkit-ai/vertexai';
+import { initializeApp } from 'firebase-admin/app';
+import { getFirestore, Timestamp } from 'firebase-admin/firestore';
+import {v4 as uuidv4} from 'uuid';
+import * as z from 'zod';
+
+// Configure Genkit with Vertex AI plugin and logging settings.
+configureGenkit({
+  plugins: [
+    vertexAI({ projectId: 'ai-startup-school-firebase', location: 'us-central1'}),
```
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts]: Changed!

File: genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    quantity: order.quantity,
+    total_sales: order.total_sales,
+    status: order.status,
+    created_at: order.created_at,
+    shipped_at: order.shipped_at,
+    delivered_at: order.delivered_at,
+    age: order.age,
+    gender: order.gender,
+    customer_rating: order.customer_rating,
+    customer_review: order.customer_review,
```
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts]: Changed. Also I tried your route. Unfortunately, it didn't work

File: genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+          // If batch size limit is reached, commit the batch and start a new one.
+          if (batchCount >= BATCH_SIZE) {
+              await batch.commit(); // Commit the batch
+              logger.log("Successfully commited a full batch of orders to Firestore.");
+              batch = db.batch();     // Create a new batch
+              batchCount = 0; 
+          }
+          i++;
+        }
+      }catch (error) {
```
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts]: Done!

File: genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+// Define constant.
+const app = initializeApp();
+const BATCH_SIZE = 20;
+
+// Create a Bone Appetit Sales Database and store it in Firestore document by document.
+export const createBoneAppetitSalesDatabase = onRequest(
+  async (request, response) => { 
+  try {
```
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts]: Adjusted to throwException instead. Realized a lot of the trycatch was for my own debugging purposes! : )

File: genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+          logger.log("Successfully added Order: " + orderObj.order_id + " to Batch.");
+          // If batch size limit is reached, commit the batch and start a new one.
+          if (batchCount >= BATCH_SIZE) {
+              await batch.commit(); // Commit the batch
+              logger.log("Successfully commited a full batch of orders to Firestore.");
+              batch = db.batch();     // Create a new batch
+              batchCount = 0; 
+          }
+          i++;
+        }
```
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/cloud-functions-version/functions/src/index.ts]: Done

File: genkit/generate-synthetic-database/cloud-functions-version/.gitignore
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/cloud-functions-version/.gitignore:1]: Changed!

File: genkit/generate-synthetic-database/README.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+| --------- | --------------------------------------------------- |
+| Author(s) | [Aiko Zhao](https://github.com/aikozhaoz) |
+
+This project leverages GenKit AI's structured output alongside the `Gemini 1.5 Pro model`, to generate synthetic sales data for a fictional dog food company called `Bone Appetit`. The generated data is then stored in a Firestore database, row by row, for further analysis or AI/ML use in applications.
+
+## Prerequisites
+### Vertex AI for LLM
+### Cloud Functions for deployment
+### Firestore for database storage
+### Firebase for application
```
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/README.md]: Changed

File: genkit/generate-synthetic-database/README.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
+# Generate Synthetic Sales Database with Genkit 
+
+|           |                                                     |
+| --------- | --------------------------------------------------- |
+| Author(s) | [Aiko Zhao](https://github.com/aikozhaoz) |
+
+This project leverages GenKit AI's structured output alongside the `Gemini 1.5 Pro model`, to generate synthetic sales data for a fictional dog food company called `Bone Appetit`. The generated data is then stored in a Firestore database, row by row, for further analysis or AI/ML use in applications.
```
[the remark this answers is not in the document]
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/README.md]: Changed!

File: genkit/generate-synthetic-database/cloud-functions/src/index.ts [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+        // If batch size limit is reached, commit the batch and start a new one.
+        if (batchCount >= BATCH_SIZE) {
+          await batch.commit(); // Commit the batch
+          logger.log(
+            "Successfully committed a full batch of orders to Firestore."
+          );
+          batch = db.batch(); // Create a new batch
+          batchCount = 0;
+        }
+        i++;
```
Comment (holtskinner) [inline genkit/generate-synthetic-database/cloud-functions/src/index.ts:289]:
```suggestion
```
This isn't needed now with the `for` loop

File: genkit/generate-synthetic-database/cloud-functions/src/index.ts [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+  async (request, response) => {
+    const generator = rateLimitedRunFlowGenerator(); // Create the generator
+    const db = getFirestore(app);
+    const collectionRef = db.collection("BoneAppetitSales");
+    let batch = db.batch();
+    let batchCount = 0;
+    for (let i = 0; i < 200; i++) {
+      const responsePromise = generator.next().value; // Get the next Promise from the generator
+      const structuredResponse = await responsePromise;
+      if (structuredResponse) {
```
Comment (holtskinner) [inline genkit/generate-synthetic-database/cloud-functions/src/index.ts:256]: I would recommend flipping this `if` statement and throwing the error here instead of doing it in the `else`. It reduces nesting and can be easier to read.

File: genkit/generate-synthetic-database/cloud-functions/src/index.ts [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+  quantity: number;
+  totalSales: number;
+  status: string;
+  createdAt: string;
+  shippedAt: string;
+  deliveredAt: string;
+  gender: string;
+  customerRating: number;
+  customerReview: string;
+}
```
Comment (holtskinner) [inline genkit/generate-synthetic-database/cloud-functions/src/index.ts:191]: It seems like a lot of this data from `Order` is replicated in multiple object types/interfaces. Would it be possible to consolidate to reduce duplication?

File: genkit/generate-synthetic-database/cloud-functions/src/index.ts [hunk 11]
```diff
@@ -0,0 +1,298 @@
...
+      );
+    } else {
+      const timeToWait = 60 * 1000 - elapsedTime;
+      yield new Promise((resolve) => setTimeout(resolve, timeToWait));
+    }
+  }
+}
+
+// Create a helper function for converting the Order object to plain object.
+function orderToPlainObject(order: Order): Record<string, any> {
```
Comment (holtskinner) [inline genkit/generate-synthetic-database/cloud-functions/src/index.ts:223]: Why exactly is a Plain object needed?

File: genkit/generate-synthetic-database/cloud-functions/src/index.ts (hunk 9 above)
Comment (aikozhaoz) [reply genkit/generate-synthetic-database/cloud-functions/src/index.ts]: Done!

File: genkit/generate-synthetic-database/cloud-functions/src/index.ts (hunk 10 above)
Comment (holtskinner) [reply genkit/generate-synthetic-database/cloud-functions/src/index.ts:191]: @aikozhaoz

**Review by aikozhaoz — COMMENTED** [review]
tldr: Genkit outputs Promise. Need to resolve Promise first. Then convert it to a plain object since Firestore only takes plan object. Also firestore only takes Timestamp not number or dates. Thats why I had to have multiple types/interfaces.

**Review by holtskinner — APPROVED** [review]
